### PR TITLE
Add CreateTxError and use as error type for TxBuilder::finish()

### LINKED ISSUE #994: `bdk::Wallet` should have context-specific error types
I agree that the errors aren't great.

However, I disagree about having a single error enum with a large number of variants. I think we should fix errors in one go with a separate PR, otherwise it's duplicate work.

_Originally posted by @evanlinjin in https://github.com/bitcoindevkit/bdk/pull/976#discussion_r1204823980_

### LINKED ISSUE #996: Wallet::get_address should return an Result
Since the persistent backend is fallible. There are `expect("TODO")` in there now which seem to have slipped through the cracks. It would be good to differentiate between those calls that are fallible and those that are infallible.
I think `AddressIndex` should go in favor of different methods that return different addresses since the error cases seem to be different.

I think with an infallible persistence backend we can have `new_address` but if it's fallible then it should be `try_new_address` and return a `Result`.

cc @notmandatory. IIRC you were thinking about how to redesign this.

### PR DESCRIPTION
**Pull request by notmandatory**
### Description

To remove some places where there were `.expect("TODO")` I added a new `CreateTxError` type which is returned from `TxBuilder::finish()`. I also updated related tests and doc tests.

Fixes https://github.com/bitcoindevkit/bdk/issues/996#issuecomment-1621036206

Also added fallible `Wallet::try_get_address()` and `try_get_internal_address()`  to return `Result` with a possible `D:WriteError` when a PersistBackend is used. This should fix bitcoindevkit/bdk#996.

I removed catch-all bdk::Error and replaced usages with new types and updated related functions, fixes bitcoindevkit/bdk#994.

### Notes to the reviewers

~~I didn't add all possible bdk::Error types that `Wallet::create_tx()` and `TxBuilder::finish()` functions might throw. It's probably not too much more work but will take a bit more research so I want to make sure this is the right general approach first.~~

I added `anyhow` to the dev-dependencies so I could remove some `.expect()` lines from the docs tests and make the examples closer to what an end user should do.  I also used the `anyhow!()` macro to replace a few places that were using the `bdk::Error::Generic` in example code.

I also moved the module level error.rs file to wallet/error.rs so no one would be tempted to make any new catch all errors and to make it clear that all the errors in it are wallet module related. 

### Changelog notice

Changed

- Updated bdk module to use new context specific error types
  - wallet: MiniscriptPsbtError, CreateTxError, BuildFeeBumpError error enums
  - coin_selection: module Error enum
- Renamed fallible Wallet address functions to try_get_address() and try_get_internal_address()

Removed

- Removed catch-all top level bdk::Error enum
- Removed impl_error macro

Added

-  Added infallible Wallet get_address(), get_internal_address functions

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### Bugfixes:

* [x] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [ ] I'm linking the issue being fixed by this PR

### CODE REVIEW

File: crates/bdk/src/wallet/mod.rs
Comment (evanlinjin) [inline crates/bdk/src/wallet/mod.rs]: I think it will be a good idea to remove the `CreateTxError::Bdk` branch in this PR as well. I think removing this will improve the API significantly (since the caller won't have to handle a large list of redundant error variants).

This can be our opportunity to avoid using `Error::Generic` and have "concrete" variants for timelock/tx-version errors.

If needed, we can have `CreateTxError` be a variant of `bdk::Error`, although I would avoid this if possible.

**Review by evanlinjin — CHANGES_REQUESTED** [review]
Thank you for taking this on!

I appreciate your attempt to be conservative in the amount of changes made. However, the current state of this PR does not result in a better API since the caller still has to handle a large list of redundant error variants.

If it was a PR to purely remove the `.expect("TODO")`s, I would just return a `bdk::Error::Generic` or introduce a new variant in `bdk::Error`.

However, I would prefer this PR to fix the error-handling API for tx creation completely (as it seems you already have done the majority of the work).

**Comment by evanlinjin** [comment]
But we probably just need a quick fix for the next release.

File: crates/bdk/src/wallet/mod.rs
Comment (notmandatory) [reply crates/bdk/src/wallet/mod.rs]: If you like the general approach of adding a new `CreateTxError` enum but would rather that I replace the `Bbk` variant with the specific errors for this PR then I'm happy to do that. I just wanted to make sure I was on the right track before chasing down all the required variants.

Comment (notmandatory) [reply crates/bdk/src/wallet/mod.rs]: I should be able to get this knocked out today or tomorrow so we don't delay the alpha.1 too much more.

**Comment by notmandatory** [comment]
I added fallible `Wallet::try_get_address()` and `try_get_internal_address()`  to return `Result` with a possible `D:WriteError` when a PersistBackend is used. This should fix bitcoindevkit/bdk#996.

**Comment by notmandatory** [comment]
I removed the `CreateTxError::Bdk` variant and moved all the needed variants from the `bdk::Error` to the `CreateTxError` enum, plus created concrete variants for any errors that were using `Generic`. A few other things I'd like to do, but don't need to be in this PR:

* add `BuildFeeBumpError` enum for `Wallet::build_fee_bump` 
* add `SignError` enum for `Wallet::sign`
* add `FinalizeError` enum for `Wallet::finalize` 
* eliminate (or greatly diminish) `bdk::Error`
* move the wallet error enums and related code to a wallet::error file
* add fallible and infallible  versions of any others Wallet functions it makes sense for

**Comment by notmandatory** [comment]
Per comment: https://github.com/bitcoindevkit/bdk/issues/1025#issuecomment-1635184287

I'm moving this to future alpha release.

File: crates/bdk/src/wallet/mod.rs
Comment (LLFourn) [inline crates/bdk/src/wallet/mod.rs:271]: This is implemented on `Wallet<()>` but it should be implemented on `Wallet<D> where D: PersistBackend<WriteError=core::convert::Infallible>`

File: example-crates/wallet_electrum/src/main.rs
Comment (LLFourn) [inline example-crates/wallet_electrum/src/main.rs]: You should not panic when this fails. It indicates that persistence failed. A good way to do this might be to use `anyhow::Context::context` and then `?` the error. It doesn't look like the examples depend on `anyhow` but they should and the main function should return `anyohw::Result`.

**Review by LLFourn — DISMISSED** [review]
+1 on adding `CreateTxError`. left some comments.

**Comment by nondiremanuel** [comment]
@notmandatory we didn't talk about this PR in the last calls. Is it correct to keep it into alpha.3 milestone? Shall we discuss it in the next call?

**Comment by notmandatory** [comment]
> @notmandatory we didn't talk about this PR in the last calls. Is it correct to keep it into alpha.3 milestone? Shall we discuss it in the next call?

I don't think we need to discuss it until after alpha.2 is released. Then I'll have to rebase it and fix a few small things and will bug folks for a re-review.

**Comment by notmandatory** [comment]
I've rebased and now need to do fixes suggested in comments.

File: crates/bdk/src/wallet/mod.rs
Comment (notmandatory) [reply crates/bdk/src/wallet/mod.rs:271]: If I do this it looks like I also have to add a `db: D` parameter to `new_no_persist()`, and then all callers (tests for now) need to specify a `db: ()`. Is there scenario you're thinking of where D is some other `Infallible` database?

**Comment by notmandatory** [comment]
This should also fix bitcoindevkit/bdk#994.  I'm in the process now of removing wallet::Error enum and replacing with specific errors. 

~~I'm also adding `thiserror` crate for less boilerplate error code.~~ We can't use `thiserror` because it isn't supported with no_std. :-(

**Comment by notmandatory** [comment]
Moved this back to draft since still have more work to do before ready for a re-review.

File: .github/workflows/cont_integration.yml
Comment (notmandatory) [inline .github/workflows/cont_integration.yml:15]: This MSRV bump should be in a separate PR once we get the go-ahead from the LDK team.

File: crates/bdk/src/wallet/mod.rs
Comment (notmandatory) [reply crates/bdk/src/wallet/mod.rs]: So this took a bit longer than a day or two to finally wrap up, but should be all done now. 😅

File: example-crates/wallet_electrum/src/main.rs
Comment (notmandatory) [reply example-crates/wallet_electrum/src/main.rs]: I added `anyhow` and removed the `.expect()` but didn't add the `.context()` because I don't think it's needed for this simple example.

**Comment by notmandatory** [comment]
This is ready for a re-review with the caveat that after CI fixes are made on `master` branch I'll need to rebase and remove the CI related commits I added here.

File: .github/workflows/cont_integration.yml
Comment (realeinherjar) [inline .github/workflows/cont_integration.yml]: Just a heads-up and to link the PRs.
This depends on #1183

File: crates/bdk/Cargo.toml
Comment (realeinherjar) [inline crates/bdk/Cargo.toml]: I am a little bit picky with changes to `Cargo.toml`.
I understand the `anyhow` addition, but why changing `env_logger` to `default-features = false`?
And I could not find this documented in the PR.

File: crates/bdk/src/descriptor/error.rs
Comment (realeinherjar) [inline crates/bdk/src/descriptor/error.rs]: What is the impact for `no-std`?
What would be the returned errors in `no-std`?

Comment (notmandatory) [reply crates/bdk/src/descriptor/error.rs]: Good question, I config'd out `fmt::Display` for errors here for consistence since it was done in other parts of the bdk package. Technically I think we could leave it and it wouldn't break `no_std` code but I believe it would make the final binary larger which could be a problem for some `no_std` embedded devices.  

@afilini any thoughts on if we should or should not include `fmt::Display` impls for errors with `no_std`?

Comment (evanlinjin) [reply crates/bdk/src/descriptor/error.rs]: I don't think it will hurt to impl Display for `no_std`. Also curious on the reasoning.

Comment (notmandatory) [reply crates/bdk/src/descriptor/error.rs]: To answer @realeinherjar there would still be returned error enums but just not formatted .to_string() or in a display string. 

I also found this list of suggestions for making small `no_std` binaries, including the recommendation to not use `core::fmt`. Looks like it only applies if someone is using `bdk` and needs an executable smaller than 20 kilobytes.

[Remove core::fmt with #![no_main] and Careful Usage of libstd](https://github.com/johnthagen/min-sized-rust#remove-corefmt-with-no_main-and-careful-usage-of-libstd)

File: .github/workflows/cont_integration.yml
Comment (notmandatory) [reply .github/workflows/cont_integration.yml]: Yup I'll remove these commits and rebase on #1183 after it's merged.

File: crates/bdk/Cargo.toml
Comment (notmandatory) [reply crates/bdk/Cargo.toml]: I originally changed this to remove one additional MSRV pinning, but was probably being a bit to aggressive so I put it back as it was. I also put it back in #1183.

File: crates/bdk/src/descriptor/error.rs
Comment (notmandatory) [reply crates/bdk/src/descriptor/error.rs]: I removed the `no_std` cfgs I added on `fmt:Display`. This change would have to be done for all bdk crates for it to have any effect on `no_std` binary size and I don't think are blocked by this. I created a new issue #1198 targeted for the 2.0 milestone.

File: crates/bdk/src/wallet/mod.rs
Comment (notmandatory) [reply crates/bdk/src/wallet/mod.rs:271]: @LLFourn I'd like to wrap up this PR, but want to make sure you're OK leaving this as it is. I don't see any reason to pass in a `db:D` parameter for `new_no_persist()` when it'd always be `()`.

**Comment by notmandatory** [comment]
I've removed the cfg for fmt::Display and created a new issue for that change (if we want to do it). Tests passing and ready for re-review.

File: crates/bdk/examples/mnemonic_to_descriptors.rs
Comment (ValuedMammal) [inline crates/bdk/examples/mnemonic_to_descriptors.rs:1]: unrelated to this PR, but the derivation path should correspond to the network, e.g. `m/86h/1h/0h/0`

File: crates/bdk/src/wallet/mod.rs
Comment (ValuedMammal) [inline crates/bdk/src/wallet/mod.rs]: nit: whitespace
```suggestion
    /// let finalized = wallet.sign(&mut psbt, SignOptions::default())?;
```

File: crates/bdk/src/wallet/tx_builder.rs
Comment (ValuedMammal) [inline crates/bdk/src/wallet/tx_builder.rs]:
```suggestion
    /// Returns a new [`Psbt`] per [BIP174].
    ///
    /// [BIP174]: https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki
```

File: crates/bdk/src/wallet/error.rs
Comment (ValuedMammal) [inline crates/bdk/src/wallet/error.rs]:
```suggestion
/// Error returned from [`TxBuilder::add_foreign_utxo`]
```

File: crates/bdk/src/wallet/mod.rs
Comment (ValuedMammal) [inline crates/bdk/src/wallet/mod.rs]: Maybe `Wallet::sign` and `Wallet::finalize_psbt` can return `Result<bool, signer::SignerError>` to avoid having to wrap it in a new type.. but that might require `signer::SignerError` to absorb an extra variant `MiniscriptPsbt`

File: crates/bdk/src/wallet/mod.rs
Comment (ValuedMammal) [inline crates/bdk/src/wallet/mod.rs:44]: should `CalculateFeeError` also implement std error so we don't have to call `expect` on `Wallet::calculate_fee`?

**Review by ValuedMammal — COMMENTED** [review]
I made minor comments, but nothing that should block this from moving forward

File: crates/bdk/src/wallet/mod.rs
Comment (notmandatory) [reply crates/bdk/src/wallet/mod.rs]: Good idea! fixed.

File: crates/bdk/src/wallet/mod.rs
Comment (notmandatory) [reply crates/bdk/src/wallet/mod.rs:44]: Makes sense, I added the implementation.

File: crates/bdk/examples/mnemonic_to_descriptors.rs
Comment (notmandatory) [reply crates/bdk/examples/mnemonic_to_descriptors.rs:1]: I created #1209 to check this in all our docs and examples.

**Comment by notmandatory** [comment]
I pushed a couple more commits to add changes suggested by @ValuedMammal. I'll squash commits to clean up the history prior to merging.

**Comment by matthiasdebernardini** [comment]
Hey! This needs fixing too, "not enough" 

https://github.com/bitcoindevkit/bdk/blob/d6a0cf0795d9f8a1550e614e96b339a18fc537dc/crates/bdk/src/descriptor/policy.rs#L524

File: crates/bdk/src/wallet/coin_selection.rs
Comment (vladimirfomene) [inline crates/bdk/src/wallet/coin_selection.rs]: Nits `use core::fmt::{self, Formatter};`

File: crates/bdk/src/wallet/mod.rs
Comment (vladimirfomene) [inline crates/bdk/src/wallet/mod.rs]: nit: add `self` on the import here to avoid importing on line 61

File: crates/bdk/src/wallet/mod.rs
Comment (vladimirfomene) [inline crates/bdk/src/wallet/mod.rs:273]: I'm curious to know why we needed to create infallible version of the address functions.

**Review by vladimirfomene — APPROVED** [review]
Just nits. LGTM. ACK 35a108b4300473277e6f575a8f781add5d555a48

File: crates/bdk/src/wallet/mod.rs
Comment (notmandatory) [reply crates/bdk/src/wallet/mod.rs:273]: I think the only reason for this is so that non-persistent / memory only applications don't need to handle a persistence error which will never happen. This is probably a rare case but I suspect there are some embedded system applications where this would be useful.

Comment (evanlinjin) [reply crates/bdk/src/wallet/mod.rs:273]: This is an excellent addition, thank you @notmandatory.

**Comment by evanlinjin** [comment]
I don't think this should be addressed in this PR, but I think we should use idiomatic naming for methods: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

I.e. `Wallet::get_address` should be `Wallet::address`.

**Review by evanlinjin — APPROVED** [review]
ACK 35a108b4300473277e6f575a8f781add5d555a48

Thank you for this work. Now the errors are more domain-specific and detailed.

File: crates/bdk/src/wallet/mod.rs
Comment (danielabrozzoni) [reply crates/bdk/src/wallet/mod.rs:271]: I think Lloyd meant that you should keep `new_no_persist` in `impl Wallet`, but move the other two methods in `impl Wallet<D: ...>`:
```diff
diff --git a/crates/bdk/src/wallet/mod.rs b/crates/bdk/src/wallet/mod.rs
index 051f9730..061a3db5 100644
--- a/crates/bdk/src/wallet/mod.rs
+++ b/crates/bdk/src/wallet/mod.rs
@@ -226,7 +226,12 @@ impl Wallet {
             NewError::Persist(_) => unreachable!("no persistence so it can't fail"),
         })
     }
+}
 
+impl<D> Wallet<D>
+where
+    D: PersistBackend<ChangeSet, WriteError = core::convert::Infallible>,
+{
     /// Infallibly return a derived address using the external descriptor, see [`AddressIndex`] for
     /// available address index selection strategies. If none of the keys in the descriptor are derivable
     /// (i.e. does not end with /*) then the same address will always be returned for any [`AddressIndex`].

```

To me, it makes sense to have `get_address` and `get_internal_address` for all the wallets that have infallible backends (which could be either `()` or some other database)

**Review by danielabrozzoni — COMMENTED** [review]
Fantastic work, than you.

I think Lloyd's comment should be fixed, as it should be fairly trivial. I left a comment on that.

**Comment by notmandatory** [comment]
> I think Lloyd's comment should be fixed, as it should be fairly trivial. I left a comment on that.

Ohhhh! 🤦‍♂️ thanks for the help here, now I get it. I'm going to push your fix, rebase, clean up the history a little, and then it should be ready to merge.

File: crates/bdk/src/wallet/mod.rs
Comment (notmandatory) [reply crates/bdk/src/wallet/mod.rs:271]: Fixed!

**Pull request merged**
